File: solutions/elixir/ocr-numbers/1/lib/ocr_numbers.ex
defmodule OcrNumbers do
  @doc """
  Given a 3 x 4 grid of pipes, underscores, and spaces, determine which number is represented, or
  whether it is garbled.
  """

  @digits [
    [" _ ", "| |", "|_|", "   "], #0
    ["   ", "  |", "  |", "   "], #1
    [" _ ", " _|", "|_ ", "   "], #2
    [" _ ", " _|", " _|", "   "], #3
    ["   ", "|_|", "  |", "   "], #4
    [" _ ", "|_ ", " _|", "   "], #5
    [" _ ", "|_ ", "|_|", "   "], #6
    [" _ ", "  |", "  |", "   "], #7
    [" _ ", "|_|", "|_|", "   "], #8
    [" _ ", "|_|", " _|", "   "], #9
  ]

  @spec convert([String.t()]) :: {:ok, String.t()} | {:error, String.t()}
  def convert(input) do
    lines = input |> Enum.chunk_every(4) # in case it's a multiple of 4
    last = List.last(lines) # validate last line - yes, I'm being lazy
    good = last |> Enum.filter(fn line -> rem(String.length(line), 3) == 0 end)
    cond do
      length(last) != 4 -> {:error, "invalid line count"}
      length(good) != 4 -> {:error, "invalid column count"}
      true ->
        {:ok, lines
          |> Enum.map(&decode_line/1)
          |> Enum.join(",")
        }
    end
  end

  defp decode_line(line) do
    Enum.map(line, fn row ->
      row
      |> String.graphemes()
      |> Enum.chunk_every(3)
      |> Enum.map(&Enum.join/1)
    end)
    # transpose
    |> List.zip
    |> Enum.map(&Tuple.to_list/1)
    # decode each cell
    |> Enum.map(&convert_one/1)
    |> Enum.join()
  end

  defp convert_one(input) do
    index = Enum.find_index(@digits, fn elem -> elem == input end)
    cond do
      index == nil -> "?"
      index >= 0 -> Integer.to_string(index)
    end
  end
end

File: solutions/typescript/ocr-numbers/1/ocr-numbers.ts

const DIGIT_MAP = {
  '0': ' _ \n| |\n|_|\n   ',
  '1': '   \n  |\n  |\n   ',
  '2': ' _ \n _|\n|_ \n   ',
  '3': ' _ \n _|\n _|\n   ',
  '4': '   \n|_|\n  |\n   ',
  '5': ' _ \n|_ \n _|\n   ',
  '6': ' _ \n|_ \n|_|\n   ',
  '7': ' _ \n  |\n  |\n   ',
  '8': ' _ \n|_|\n|_|\n   ',
  '9': ' _ \n|_|\n _|\n   ',
}

export function convert(grid: string) {
  /* 
Given a 3 x 4 grid of pipes, underscores, and spaces, determine which number is represented, or whether it is garbled.
*/
  /* 
  split the grid into 3x4 chunks
  */
  // Split into rows
  const rows = grid.split('\n')
  
  // Calculate number of digits (each is 3 chars wide)
  const numDigits = rows[0].length / 3

  // For each group of 4 lines
  const groups: string[][] = [];
  for (let group = 0; group < rows.length; group += 4) {
    const groupChunks: string[] = [];
    // For each digit position in this group
    for (let i = 0; i < numDigits; i++) {
      const digit = [
        rows[group].slice(i * 3, (i + 1) * 3),
        rows[group + 1].slice(i * 3, (i + 1) * 3), 
        rows[group + 2].slice(i * 3, (i + 1) * 3),
        rows[group + 3].slice(i * 3, (i + 1) * 3)
      ].join('\n');
      groupChunks.push(digit);
    }
    groups.push(groupChunks);
  }

  const ocr_digits = Object.values(DIGIT_MAP);
  const ans_digits = Object.keys(DIGIT_MAP);
  return groups.map(line => {
    return line.map(digit => {
      const index = ocr_digits.findIndex(d => d === digit);
      if (index === -1) return '?';
      return ans_digits[index];
    }).join('');
  }).join(',');
}
